python/kserve/kserve/protocol/dataplane.py

```diff
@@ -217,10 +216,28 @@ def decode(self, body, headers) -> Union[Dict, InferRequest]:
                     body = orjson.loads(body)
                 except orjson.JSONDecodeError as e:
                     raise InvalidInput(f"Unrecognized request format: {e}")
-        t2 = time.time()
-        logging.debug(f"decoded request in {round((t2 - t1) * 1000, 9)}ms")
         return body
 
+    def decode_cloudevent(self, payload):
+        response = payload
+        if isinstance(payload, CloudEvent):
+            # Try to decode and parse JSON UTF-8 if possible, otherwise
+            # just pass the CloudEvent data on to the predict function.
+            # This is for the cases that CloudEvent encoding is protobuf, avro etc.
+            try:
+                response = orjson.loads(response.data.decode('UTF-8'))
+            except (orjson.JSONDecodeError, UnicodeDecodeError) as e:
+                # If decoding or parsing failed, check if it was supposed to be JSON UTF-8
+                if "content-type" in payload._attributes and \
+                        (payload._attributes["content-type"] == "application/cloudevents+json" or
+                         payload._attributes["content-type"] == "application/json"):
+                    raise InvalidInput(f"Failed to decode or parse binary json cloudevent: {e}")
+
+        elif isinstance(payload, dict):
+            if is_structured_cloudevent(payload):
+                response = payload["data"]
+        return response
```

Comment:
I think this block of code can be merged into the `decode` function.
```python
if headers and has_binary_headers(headers):
    payload = self.get_binary_cloudevent(body, headers)
    try:
        body = orjson.loads(payload.data.decode('UTF-8'))
    except (orjson.JSONDecodeError, UnicodeDecodeError) as e:
         # If decoding or parsing failed, check if it was supposed to be JSON UTF-8
         if "content-type" in payload._attributes and \
                        (payload._attributes["content-type"] == "application/cloudevents+json" or
                         payload._attributes["content-type"] == "application/json"):
             raise InvalidInput(f"Failed to decode or parse binary json cloudevent: {e}")
elif isinstance(body, dict) and is_structured_cloudevent(payload):
    body = body["data"]
elif type(body) is bytes:
    try:
        body = orjson.loads(body)
    except orjson.JSONDecodeError as e:
        raise InvalidInput(f"Unrecognized request format: {e}")
```

Reply:
[create_response_cloudevent](https://github.com/kserve/kserve/blob/9bd6375957a82db022491daac49e64fdde604ac0/python/kserve/kserve/utils/utils.py#L93) requires cloudevent `attributes` field. This function is used in [encode method](https://github.com/kserve/kserve/blob/9bd6375957a82db022491daac49e64fdde604ac0/python/kserve/kserve/protocol/dataplane.py#L236). if we add this logic to decode method `attributes` field will not be available because the returned body only contains cloudevent `data`.

Reply:
I see that’s the binary cloudevent case, can we return a tuple with the attributes and then pass along to the encode function ?

Reply:
done